### PR TITLE
[⚠️ removed a default argument] Make `AttentionMaskConverter` compatible with `torch.compile(..., fullgraph=True)`

### LINKED ISSUE #27789: _prepare_4d_causal_attention_mask doesn't work with torch.compile
### System Info

Transformers version: 4.35.2
Pytorch version: 2.1.1+cu118
Python version: 3.11

### Who can help?

Tagging @ArthurZucker as the original author

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Code to reproduce
```python
import torch
import torch.nn as nn
from transformers.modeling_attn_mask_utils import _prepare_4d_causal_attention_mask

# LLama models use this wrapper but this does not fix the compilation issue.
# if is_torch_fx_available():
#    _prepare_4d_causal_attention_mask = torch.fx.wrap(_prepare_4d_causal_attention_mask)

class Model(nn.Module):

    def forward(self, inputs_embeds):
        batch_size, seq_length, _ = inputs_embeds.shape
        past_key_values_length = 10
        attention_mask = _prepare_4d_causal_attention_mask(
            None, (batch_size, seq_length), inputs_embeds, past_key_values_length
        )
        return attention_mask

model = Model()
model = torch.compile(model, fullgraph=True)
model(torch.ones([1,5, 32]))
```
Gives a torch.compile error:
```
[...]
    return SourcelessBuilder()(tx, val).add_options(options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/.virtualenvs/pytorch311/lib/python3.11/site-packages/torch/_dynamo/variables/builder.py", line 1650, in __call__
    unimplemented(f"Unexpected type in sourceless builder {type(value)}")
  File "/home/ubuntu/.virtualenvs/pytorch311/lib/python3.11/site-packages/torch/_dynamo/exc.py", line 172, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: Unexpected type in sourceless builder <class 'torch.dtype'>

from user code:
   File "<stdin>", line 6, in forward
  File "/home/ubuntu/.virtualenvs/pytorch311/lib/python3.11/site-packages/transformers/modeling_attn_mask_utils.py", line 197, in _prepare_4d_causal_attention_mask
    attention_mask = attn_mask_converter.to_causal_4d(
```

Note that I could reproduced this by e.g. trying to compile a Llama2 model.

Removing the dtype argument and just setting dtype=torch.float32 in the function body compiles.
I'm not sure whether this is a bug in the torch compiler or unintended usage in transformers, so happy to open an issue there as well if required :)

### Expected behavior

No errors during torch.compile stge.
Let me know if there's something I can do to help debug!

### PR DESCRIPTION
**Pull request by fxmarty**
As per title, fixes https://github.com/huggingface/transformers/issues/27789.

This issue is only for PyTorch 2.1 and has been fixed in torch nightly.

### CODE REVIEW

File: src/transformers/modeling_attn_mask_utils.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ def to_causal_4d(
         batch_size: int,
         query_length: int,
         key_value_length: int,
-        dtype: torch.dtype = torch.float32,
+        dtype: torch.dtype,
```
Comment (fxmarty) [inline src/transformers/modeling_attn_mask_utils.py:69]: It is fine to remove the default (that is the cause of the error), as in `_prepare_4d_causal_attention_mask`, `_prepare_4d_attention_mask` and `_create_4d_causal_attention_mask` privately exposed methods `dtype` is always passed.

Comment (ArthurZucker) [reply src/transformers/modeling_attn_mask_utils.py:69]: But to_causal_4d is part of the public api no? O this is breaking 😅

**Review by ArthurZucker — COMMENTED** [review]
I think it should be alright even if breaking as it's mostly used for internal calls and has not really been here for a long time. Could you link the pytorch issue?

**Review by ArthurZucker — APPROVED** [review]
Alright with me let's just add a ⚠️

**Pull request merged**